rust/connlib/tunnel/src/dns.rs

```diff
@@ -226,6 +246,7 @@ impl StubResolver {
             )));
         }
 
+        // `match_resource` is `O(N)` which we deem fine for DNS queries.
```

Comment:
@conectado Do you agree with this? I thought a bit on how to make this more efficient but I am not sure it is worth it?

Reverse-lookup is already optimised because we store the resource ID along with the assigned IP address for a certain DNS resource.

What we could do here is to store a tree of patterns instead, like:

- `com`
	- `example`
		- `**.example.com`
	- `domain`
- `net`
	- `example`	

And then upon matching, we can iterate through the labels of the domain and thus avoid matching against all of them. I've not benchmarked it but even matching against 100 DNS resources (which is a lot I guess?) should be very quick.

@jamilbk Do we have any data on the number of DNS resources that customers use?

Reply:
O(N) should be fine for the foreseeable future. When we reach 1000 resources we can use a search tree like you suggest, but this is only for lookups and not packet matching correct?

Reply:
> O(N) should be fine for the foreseeable future. When we reach 1000 resources we can use a search tree like you suggest, but this is only for lookups and not packet matching correct?

Yes, this code will run once for every DNS request we receive, not during packet routing.

Reply:
In that case, it runs even for non-resources right? To determine if they match?

Yeah it might be good to invest a little extra time to make this O(log N) just so we're done with it. If this becomes an issue it would likely become a big issue I imagine and might surprise us/customer in a bad way.

Reply:
I notice during a typical web browser session Chrome can send upwards of 100 queries a second, so I imagine on mobile for 1,000 Resources let's say, that might eat a little more power than we need to.

Reply:
> In that case, it runs even for non-resources right? To determine if they match?
> 
> Yeah it might be good to invest a little extra time to make this O(log N) just so we're done with it. If this becomes an issue it would likely become a big issue I imagine and might surprise us/customer in a bad way.

Yep, non resources will also trigger this. I'll do some benchmarking.

Reply:
```
firezone_tunnel               fastest       │ slowest       │ median        │ mean          │ samples │ iters
╰─ dns                                      │               │               │               │         │
   ╰─ benches                               │               │               │               │         │
      ╰─ match_domain_linear                │               │               │               │         │
         ├─ 10                179.7 ns      │ 1.141 µs      │ 259.7 ns      │ 265.6 ns      │ 100     │ 100
         ├─ 100               270.2 ns      │ 608.5 ns      │ 460.6 ns      │ 461.5 ns      │ 100     │ 400
         ├─ 1000              420.7 ns      │ 17.46 µs      │ 2.904 µs      │ 3.334 µs      │ 100     │ 100
         ├─ 10000             550.7 ns      │ 142.1 µs      │ 41.1 µs       │ 40.71 µs      │ 100     │ 100
         ╰─ 100000            5.129 µs      │ 7.812 ms      │ 1.631 ms      │ 1.817 ms      │ 100     │ 100
```

These are the results on my computer for the linear search. Is it worth optimising? We are within network latency here for answering DNS requests, even if the user has 100_000 DNS resources configured.

Reply:
@jamilbk @conectado In my eyes, this performance is good enough but can also invest time into optimising this if you'd like.

Note that I am not using any wildcards in these tests so I think it represents the computationally worst-case where each character in each domain needs to be compared. I haven't closely analysed the algorithms in `glob` though.

Reply:
@thomaseizinger Thanks for providing those benchmarks. I am just now catching up on them.

Considering DNS is typically across a network, I think this is fine.